### PR TITLE
Segwit update - tx, test_tx, helpers(P2WPKH) updated for BIP143; SPQR sighash-checking script; simple locktime fix

### PR DESCRIPTION
**Pull request by drmoog**
Take a look at this with the update helper values, if you would

### CODE REVIEW

**Comment by drmoog** [comment]
(doesn't pass tox yet)

**Comment by drmoog** [comment]
Update: BIP143 is now live in tx, test_tx, helpers -- but a dependency in simple.py is now borken. I'm on it.

**Comment by drmoog** [comment]
Update: All clear, all passes tox on my end, also fixed locktime in simple.py and its tests

File: SPQR.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# # Generates sighash values from available information on a raw transaction,
```
Comment (prestwich) [inline SPQR.py]: Can we `git mv` this into the tests directory? Maybe make `riemann/tests/scripts`

File: riemann/simple.py [hunk 2]
```diff
@@ -326,7 +326,7 @@ def legacy_tx(tx_ins, tx_outs, **kwargs):
                        if 'joinsplit_sig' in kwargs else None))
 
 
-def witness_tx(tx_ins, tx_outs, tx_witnesses):
+def witness_tx(tx_ins, tx_outs, tx_witnesses, **kwargs):
     '''
     list(TxIn), list(TxOut), list(InputWitness) -> Tx
```
Comment (prestwich) [inline riemann/simple.py]: Please update this docstring to match format of others in the file :+1:

File: riemann/tests/helpers.py [hunk 3]
```diff
@@ -276,23 +276,23 @@
...
+            'locktime': bytes.fromhex('cafd0700'),
             'ins': [
                 {
                     'id': 0,
                     'hash': bytes.fromhex('d2941b532f6d3d54d596345b50972b3995983239884037a52aab799ec84292ee'),
                     'index': bytes.fromhex('00000000'),
-                    'value': bytes.fromhex(''),
-                    'pk_script': bytes.fromhex('0014758ce550380d964051086798d6546bebdca27a73'),
+                    'value': bytes.fromhex('c0d4010000000000'),
+                    'pk_script': bytes.fromhex('a914419fbecd4b8f90c779b773a0a9010dd25423b06b87'),
```
Comment (prestwich) [inline riemann/tests/helpers.py]: Correct this back to `0014758ce550380d964051086798d6546bebdca27a73`

File: riemann/tx/tx.py [hunk 4]
```diff
@@ -712,6 +720,14 @@ def sighash_single(self, index, script=None,
                                         sighash_type=SIGHASH_SINGLE,
                                         anyone_can_pay=anyone_can_pay)
 
+        if self.flag is not None:
```
Comment (prestwich) [inline riemann/tx/tx.py]: replace these with `if self.is_witness()`

File: riemann/tx/tx.py [hunk 5]
```diff
@@ -683,6 +683,14 @@ def sighash_all(self, index, script=None,
                                         sighash_type=SIGHASH_ALL,
                                         anyone_can_pay=anyone_can_pay)
 
+        if self.flag is not None:
```
Comment (prestwich) [inline riemann/tx/tx.py]: replace these with `if self.is_witness()`

File: riemann/tx/tx.py [hunk 6]
```diff
@@ -735,6 +751,52 @@ def sighash_single(self, index, script=None,
...
         return self._sighash_final_hashing(copy_tx, SIGHASH_SINGLE)
 
+    def segwit_sighash(self, index, script, prevout_value=None,
+                       sighash_type=None, anyone_can_pay=False):
+        '''
+        this function sets up sighash in BIP143 style
+        https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+        https://ricette.giallozafferano.it/Spaghetti-alla-Norma.html
+        '''
+        # self.validate_bytes(prevout_value, 8)
```
Comment (prestwich) [inline riemann/tx/tx.py]: Uncomment, or delete if unnecessary

File: riemann/tx/tx.py [hunk 7]
```diff
@@ -834,6 +897,13 @@ def _adjusted_sighash_type(self, sighash_type, anyone_can_pay):
             sighash = sighash | SIGHASH_ANYONECANPAY
         return utils.i2le_padded(sighash, 4)
 
+    def _segwit_sighash_adjust(self, sighash_type, anyone_can_pay):
```
Comment (prestwich) [inline riemann/tx/tx.py]: Can we rename `_adjusted_sighash_type` to `_forkid_sighash_adjustment`

File: riemann/tx/tx.py [hunk 8]
```diff
@@ -834,6 +897,13 @@ def _adjusted_sighash_type(self, sighash_type, anyone_can_pay):
...
+        # sighash type altered to include ANYONECANPAY
+        sighash = sighash_type
```
Comment (prestwich) [inline riemann/tx/tx.py]:
```
    if anyone_can_pay:
        sighash_type = sighash_type | SIGHASH_ANYONECANPAY
    return utils.i2le_padded(sighash_type, 4)
````

File: riemann/tests/helpers.py (hunk 3 above)
Comment (drmoog) [reply riemann/tests/helpers.py]: (ouo) b

**Comment by drmoog** [comment]
Have imbued it with the power of the requested changes

**Pull request merged**